### DISCUSSION TITLE
dispose method in ConsumerWidget?

### DISCUSSION THREAD

**Discussion by kennethjyang** [rrousselGit/riverpod#2828] [Q&A] [answered]
I have a `TextEditingController` which should be disposed of in the `dispose` method of my widget. Typically, this is done by overriding the `dispose` method of a stateful widget, but I'm using a `ConsumerWidget` which does not seem to have this function available to override.

How do I properly dispose of my `TextEditingController` after this widget exits? Or will it be cleaned up automatically?

**Comment by AhmedLSayed9** [comment, marked as answer]
It'll not be cleaned up automatically. You can use `ConsumerStatefulWidget` which act as `StatefulWidget` + the ability to read providers.

A better option is to use [useTextEditingController](https://pub.dev/documentation/flutter_hooks/latest/flutter_hooks/useTextEditingController-constant.html) from Hooks, this one will automatically handle the dispose. For using Hooks+Riverpod you should extend [HookConsumerWidget](https://pub.dev/documentation/hooks_riverpod/latest/hooks_riverpod/HookConsumerWidget-class.html).

**Reply by kennethjyang** [reply]
Thanks!

**Reply by kennethjyang** [reply]
Is there a version of this for custom TextEditingControllers? That is, a class that extends `TextEditingController`?

**Reply by AhmedLSayed9** [reply]
Check https://pub.dev/packages/flutter_hooks#how-to-create-a-hook
